src/charm_state.py

```diff
@@ -114,7 +240,109 @@ def from_charm(cls, charm: CharmBase) -> "CharmConfig":
+        )
+
+    @root_validator
+    @classmethod
+    def check_fields(cls, values: dict) -> dict:
+        """Validate the general charm configuration."""
+        reconcile_interval = values.get("reconcile_interval")
+        # By property definition, this cannot be None. Needed for mypy type check.
+        assert reconcile_interval is not None, "Unreachable code"  # nosec for [B101:assert_used]
+
+        if reconcile_interval < 2:
+            logger.exception("The virtual-machines configuration must be int")
+            raise ValueError(
+                "The reconcile_interval configuration needs to be greater or equal to 2"
+            )
+
+        return values
+
+
+class RunnerCharmConfig(BaseModel):
+    """Runner configurations for the charm.
+
+    Attributes:
+        virtual_machines: Number of virtual machine-based runner to spawn.
+        virtual_machine_resources: Hardware resource use by one virtual machine for a runner.
+        runner_storage: Storage to be used as disk for the runner.
+    """
+
+    virtual_machines: int
+    virtual_machine_resources: VirtualMachineResources
+    runner_storage: RunnerStorage
+
+    @classmethod
+    def from_charm(cls, charm: CharmBase) -> "RunnerCharmConfig":
+        """Initialize the config from charm.
+
+        Args:
+            charm: The charm instance.
+
+        Returns:
+            Current config of the charm.
+        """
+        try:
+            runner_storage = RunnerStorage(charm.config["runner-storage"])
+        except ValueError as err:
+            raise CharmConfigInvalidError("Invalid runner-storage configuration") from err
+
+        try:
+            virtual_machines = int(charm.config["virtual-machines"])
+        except ValueError as err:
+            raise CharmConfigInvalidError(
+                "The virtual-machines configuration must be int"
+            ) from err
+
+        try:
+            cpu = int(charm.config["vm-cpu"])
+        except ValueError as err:
+            raise CharmConfigInvalidError("Invalid vm-cpu configuration") from err
+
+        virtual_machine_resources = VirtualMachineResources(
+            cpu, charm.config["vm-memory"], charm.config["vm-disk"]
+        )
+
+        return cls(
+            virtual_machines=virtual_machines,
+            virtual_machine_resources=virtual_machine_resources,
+            runner_storage=runner_storage,
+        )
+
+    @root_validator
+    @classmethod
+    def check_fields(cls, values: dict) -> dict:
+        """Validate the runner configuration."""
+        virtual_machines = values.get("virtual_machines")
+        resources = values.get("virtual_machine_resources")
+        # By property definition, this cannot be None. Needed for mypy type check.
+        assert virtual_machines is not None, "Unreachable code"  # nosec for [B101:assert_used]
+        assert resources is not None, "Unreachable code"  # nosec for [B101:assert_used]
+
```

Comment:
what about a `cast` from `typing` if it is primarily for mypy?

Reply:
Thanks, it worked.